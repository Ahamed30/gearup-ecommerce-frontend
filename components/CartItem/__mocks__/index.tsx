export const cartItemData = {
  id: "clk8dhn0s0udx0bpbmqrkru3f",
  quantity: 1,
  heroImage: {
    id: "clk8didmw0ug60bpbkp00pcrv",
    url: "https://media.graphassets.com/tEoKFf1JQ1a2QsVIOTdH",
  },
  productName: "ADIDAS 4DFWD X PARLEY RUNNING SHOES",
  price: 125,
  salePrice: 100,
  size: "10",
  color: "Yellow",
};
