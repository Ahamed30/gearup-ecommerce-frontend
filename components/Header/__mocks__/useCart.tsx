// __mocks__/useCart.ts
import { createContext, useContext } from "react";

// Create a mock cart context with a default value of null (no cart data).
const CartContext = createContext(null);

// Create a custom hook to access the mock cart context.
export const useCart = () => {
  const cart = useContext(CartContext);
  if (cart === null) {
    throw new Error("useCart must be used within a CartProvider");
  }
  return cart;
};
