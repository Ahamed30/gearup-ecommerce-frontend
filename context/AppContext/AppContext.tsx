import { type ReactNode, useContext, useState, createContext } from "react";

export interface AppContextType {
  isLoading: boolean;
  // eslint-disable-next-line no-unused-vars
  setIsLoading: (isLoading: boolean) => void;
}

//@ts-expect-error: ignore initial context creation
const AppContext = createContext<AppContextType>(null);

export const useApp = () => useContext(AppContext);

export interface AppContextProviderProps {
  children: ReactNode;
}
export const AppContextProvider = ({ children }: AppContextProviderProps) => {
  const [isLoading, setIsLoading] = useState(false);

  return (
    <AppContext.Provider value={{ isLoading, setIsLoading }}>
      {children}
    </AppContext.Provider>
  );
};
