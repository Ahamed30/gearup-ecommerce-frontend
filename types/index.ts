export * from "./header";
export * from "./link";
export * from "./footer";
export * from "./linkColumn";
export * from "./subscribeForm";
export * from "./asset";
export * from "./homePage";
export * from "./productCard";
