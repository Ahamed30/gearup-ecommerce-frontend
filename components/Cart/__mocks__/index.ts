import { CartType } from "@/types";

export const cartWithItems: CartType = {
  items: [
    {
      id: "clk8dhn0s0udx0bpbmqrkru3f",
      quantity: 1,
      heroImage: {
        id: "clk8didmw0ug60bpbkp00pcrv",
        url: "https://media.graphassets.com/tEoKFf1JQ1a2QsVIOTdH",
      },
      productName: "ADIDAS 4DFWD X PARLEY RUNNING SHOES",
      price: 125,
      salePrice: 100,
      size: "10",
      color: "Yellow",
    },
    {
      id: "clk29a3wv2o250bpqy9fxnqxq",
      quantity: 1,
      heroImage: {
        id: "clk29209r2oap0bnxkd7aqmi3",
        url: "https://media.graphassets.com/QB2jGvWSqGxI4J0IZMpj",
      },
      productName: "Nike Air Max Plus III",
      price: 170,
      salePrice: 120,
      size: "8",
      color: "Blue",
    },
  ],
  itemCount: 2,
  cartSubTotal: 270,
};
