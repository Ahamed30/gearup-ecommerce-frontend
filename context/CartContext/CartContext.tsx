import { createContext, useContext, useState } from "react";
import type { ReactNode, Key } from "react";
import { Cart, CartItemType } from "@/types";
import { getCookie, removeCookie, setCookie } from "@/utils/cookie";

interface CartContextType {
  addToCart: (item: CartItemType) => void; // when updating via backend these can be update to Promise<void> from void
  cart: Cart | null;
  // clearCartAlertState: () => void;
  removeFromCart: (itemId: Key, size: string) => void;
  updateCartItem: (itemId: string, updatedItem: CartItemType) => void;
  updateCart: (cart: Cart) => void;
  emptyCart: () => void;
}

//@ts-expect-error: ignore initial context creation
const CartContext = createContext<CartContextType>(null);

export const useCart = () => useContext(CartContext);

export interface CartContextProviderProps {
  cart?: Cart | null;
  children: ReactNode;
}

export const CartContextProvider = ({
  cart: initialCart = null,
  children,
}: CartContextProviderProps) => {
  const cartFromCookie: Cart | null = getCookie("cart");
  // console.log("...", cartFromCookie);
  const [cart, setCart] = useState(initialCart ?? cartFromCookie);

  const addToCart = (item: CartItemType) => {
    let isItemExists = false;
    if (cart?.items) {
      cart.items.forEach((element) => {
        if (element.id === item.id && element.size === item.size) {
          cart.cartSubTotal =
            (cart.cartSubTotal ?? 0) + (item?.salePrice ?? item.price);
          element.quantity += 1;
          isItemExists = true;
          setCookie("cart", cart);
          return;
        }
      });
    }
    if (isItemExists) return;
    setCart((prevCart) => {
      const updatedCart: Cart = { ...prevCart };

      updatedCart.items = [...(prevCart?.items || []), item];
      updatedCart.itemCount = (updatedCart.itemCount ?? 0) + 1;
      updatedCart.cartSubTotal =
        (updatedCart.cartSubTotal ?? 0) + (item?.salePrice ?? item.price);
      setCookie("cart", updatedCart);
      return updatedCart;
    });
  };

  const removeFromCart = (itemId: Key, size: string) => {
    setCart((prevCart) => {
      const updatedCart: Cart = { ...prevCart };
      updatedCart.items = updatedCart.items?.filter((item) => {
        if (item.id === itemId && item.size === size) {
          updatedCart.cartSubTotal =
            (updatedCart.cartSubTotal ?? 0) -
            (item?.salePrice ?? item.price) * item.quantity;
          return false;
        }
        return true;
      });

      updatedCart.itemCount = (updatedCart.itemCount ?? 1) - 1;
      setCookie("cart", updatedCart);
      return updatedCart;
    });
    console.log("-->", cart);
  };

  const updateCartItem = (itemId: Key, updatedItem: CartItemType) => {
    setCart((prevCart) => {
      const updatedCart: Cart = { ...prevCart };

      updatedCart.items = updatedCart.items?.map((item) =>
        item.id === itemId ? updatedItem : item
      );
      setCookie("cart", updatedCart);
      return updatedCart;
    });
  };

  const updateCart = (cart: Cart) => {
    setCart(cart);
    setCookie("cart", cart);
  };

  const emptyCart = () => {
    removeCookie("cart");
    setCart(null);
  };

  return (
    <CartContext.Provider
      value={{
        cart,
        addToCart,
        removeFromCart,
        updateCartItem,
        updateCart,
        emptyCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};
