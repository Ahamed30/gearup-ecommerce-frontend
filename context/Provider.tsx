"use client";

import { SessionProvider } from "next-auth/react";
import type { ReactNode } from "react";
import { AppContextProvider } from "./AppContext";
import { CartContextProvider } from "./CartContext";
import { UserContextProvider } from "./UserContext";

interface ProviderProps {
  children: ReactNode;
}

export const Provider = ({ children }: ProviderProps) => {
  return (
    <SessionProvider>
      <AppContextProvider>
        <UserContextProvider>
          <CartContextProvider>{children}</CartContextProvider>
        </UserContextProvider>
      </AppContextProvider>
    </SessionProvider>
  );
};
