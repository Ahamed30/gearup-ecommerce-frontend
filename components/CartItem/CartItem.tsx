import { useState } from "react";
import Image from "next/image";
import { useRouter } from "next/navigation";
import { useCart } from "@/context/CartContext";
import { Typography } from "../Typography";
import { CartItemType } from "@/types";
import { Modal } from "../Modal";

interface CartItemProps {
  product: CartItemType;
}

export const CartItem = ({ product }: CartItemProps) => {
  const router = useRouter();
  const { removeFromCart } = useCart();
  const [isFavourite, setIsFavouriteSelected] = useState<boolean>(false);
  const [showModal, setShowModal] = useState<boolean>(false);

  const {
    id,
    productName,
    heroImage,
    size,
    quantity,
    price,
    salePrice,
    color,
  } = product;

  const removeItem = () => {
    removeFromCart(id, size);
  };

  return (
    <div role="cartItem">
      <div className="flex gap-[16px] md:gap-[60px] lg:gap-[24px] mb-[16px]">
        <div className="relative w-1/2 lg:w-1/4 h-[200px] lg:h-[180px]">
          <Image
            src={`${heroImage.url}`}
            fill
            className="object-cover rounded-[24px]"
            alt="ProductImage"
          />
        </div>
        <div className="w-3/4 flex-grow">
          <div className="lg:flex lg:justify-between gap-5">
            <div className="w-full">
              <Typography
                variant="headline"
                className="text-base lg:text-xl font-semibold mb-[8px] uppercase"
              >
                {productName}
              </Typography>
              {color && (
                <Typography className="text-sm lg:text-lg font-semibold mb-[8px]">
                  {color}
                </Typography>
              )}
              <div className="text-sm lg:w-full lg:text-lg font-semibold lg:flex  gap-10 mb-[8px] lg:mb-[32px]">
                <div className="flex items-center gap-2">
                  <Typography className="text-base lg:text-xl">
                    Size {size}
                  </Typography>
                </div>
                <div className="flex items-center gap-2">
                  <Typography className="text-base lg:text-xl">
                    Quantity {quantity}
                  </Typography>
                </div>
                <div className="flex items-center gap-2">
                  <Typography
                    className="text-base lg:text-xl underline underline-offset-2 cursor-pointer"
                    onClick={() => router.push(`/product/${id}`)}
                  >
                    Edit
                  </Typography>
                </div>
              </div>
            </div>
            <div className="flex flex-row lg:flex-col mb-[8px] lg:mb-[32px] gap-3">
              {salePrice && (
                <Typography
                  variant="headline"
                  color="#EF4444"
                  className="text-xl lg:text-2xl font-semibold"
                >
                  ${salePrice}
                </Typography>
              )}
              <Typography
                variant="headline"
                color="#4A69E2"
                className="text-xl lg:text-2xl font-semibold"
              >
                {salePrice ? <s>${price}</s> : `$${price}`}
              </Typography>
            </div>
          </div>

          <div className="flex gap-[24px]">
            <Image
              src={isFavourite ? "/heart-red.svg" : "/heart-black.svg"}
              height={32}
              width={32}
              alt="Add to favourite"
              className="cursor-pointer"
              onClick={() => setIsFavouriteSelected(!isFavourite)}
            />
            <Image
              src="/bin.svg"
              height={32}
              width={32}
              alt="Delete Item"
              className="cursor-pointer"
              onClick={() => setShowModal(true)}
            />
          </div>
        </div>
      </div>
      {showModal && (
        <Modal
          title="Remove Product"
          subText="Are you sure you want to remove this product?"
          setShowModal={setShowModal}
          handleClick={removeItem}
        />
      )}
    </div>
  );
};
