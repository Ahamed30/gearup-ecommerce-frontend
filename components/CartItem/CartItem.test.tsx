import { fireEvent, render, screen } from "@testing-library/react";
import { AppContextProvider } from "@/context/AppContext";
import { CartContextProvider } from "@/context/CartContext";
import { CartItemType } from "@/types";
import { cartItemData } from "./__mocks__";
import { CartItem } from "./CartItem";

const mockRouterPush = jest.fn();
jest.mock("next/navigation", () => ({
  useRouter: () => ({
    push: mockRouterPush,
  }),
}));

const renderComponent = (cartItem: CartItemType = cartItemData) => {
  return render(
    <AppContextProvider>
      <CartContextProvider>
        <CartItem product={cartItem} />
      </CartContextProvider>
    </AppContextProvider>
  );
};

describe("Cart", () => {
  test("should match the snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("should render the cart item with all necessary information", () => {
    renderComponent();
    const cartItemElement = screen.getByRole("cartItem");
    expect(cartItemElement).toBeInTheDocument();
    const productNameElement = screen.getByText(
      "ADIDAS 4DFWD X PARLEY RUNNING SHOES"
    );
    expect(productNameElement).toBeInTheDocument();
    const sizeElement = screen.getByText("Size 10");
    expect(sizeElement).toBeInTheDocument();
    const quantitylement = screen.getByText("Quantity 1");
    expect(quantitylement).toBeInTheDocument();
    const priceElement = screen.getByText("$125");
    expect(priceElement).toBeInTheDocument();
    const salePriceElement = screen.getByText("$100");
    expect(salePriceElement).toBeInTheDocument();
    const colorElement = screen.getByText("Yellow");
    expect(colorElement).toBeInTheDocument();
  });

  test("should toggle favourite state when heart icon is clicked", () => {
    renderComponent();

    const favouriteButton = screen.getByAltText("Add to favourite");

    fireEvent.click(favouriteButton);

    expect(favouriteButton.getAttribute("src")).toContain("heart-red.svg");

    fireEvent.click(favouriteButton);

    expect(favouriteButton.getAttribute("src")).toContain("heart-black.svg");
  });

  test("should open the modal when delete icon is clicked", () => {
    renderComponent();
    const deleteButton = screen.getByAltText("Delete Item");

    fireEvent.click(deleteButton);

    const dialogBox = screen.getByRole("dialog");
    expect(dialogBox).toBeInTheDocument();
  });

  test("should close the modal when No is clicked in the modal", () => {
    renderComponent();
    const deleteButton = screen.getByAltText("Delete Item");
    fireEvent.click(deleteButton);
    const dialogBox = screen.getByRole("dialog");
    expect(dialogBox).toBeInTheDocument();
    fireEvent.click(screen.getByText("No, cancel"));
    expect(dialogBox).not.toBeInTheDocument();
  });
});
