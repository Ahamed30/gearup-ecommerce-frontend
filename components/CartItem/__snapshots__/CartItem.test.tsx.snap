// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart should match the snapshot 1`] = `
<div>
  <div
    role="cartItem"
  >
    <div
      class="flex gap-[16px] md:gap-[60px] lg:gap-[24px] mb-[16px]"
    >
      <div
        class="relative w-1/2 lg:w-1/4 h-[200px] lg:h-[180px]"
      >
        <img
          alt="ProductImage"
          class="object-cover rounded-[24px]"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="100vw"
          src="/_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=3840&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fmedia.graphassets.com%2FtEoKFf1JQ1a2QsVIOTdH&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
        />
      </div>
      <div
        class="w-3/4 flex-grow"
      >
        <div
          class="lg:flex lg:justify-between gap-5"
        >
          <div
            class="w-full"
          >
            <p
              class="className text-base lg:text-xl font-semibold mb-[8px] uppercase"
              style="color: rgb(35, 35, 33);"
            >
              ADIDAS 4DFWD X PARLEY RUNNING SHOES
            </p>
            <p
              class="className text-sm lg:text-lg font-semibold mb-[8px]"
              style="color: rgb(35, 35, 33);"
            >
              Yellow
            </p>
            <div
              class="text-sm lg:w-full lg:text-lg font-semibold lg:flex  gap-10 mb-[8px] lg:mb-[32px]"
            >
              <div
                class="flex items-center gap-2"
              >
                <p
                  class="className text-base lg:text-xl"
                  style="color: rgb(35, 35, 33);"
                >
                  Size 
                  10
                </p>
              </div>
              <div
                class="flex items-center gap-2"
              >
                <p
                  class="className text-base lg:text-xl"
                  style="color: rgb(35, 35, 33);"
                >
                  Quantity 
                  1
                </p>
              </div>
              <div
                class="flex items-center gap-2"
              >
                <p
                  class="className text-base lg:text-xl underline underline-offset-2 cursor-pointer"
                  style="color: rgb(35, 35, 33);"
                >
                  Edit
                </p>
              </div>
            </div>
          </div>
          <div
            class="flex flex-row lg:flex-col mb-[8px] lg:mb-[32px] gap-3"
          >
            <p
              class="className text-xl lg:text-2xl font-semibold"
              style="color: rgb(239, 68, 68);"
            >
              $
              100
            </p>
            <p
              class="className text-xl lg:text-2xl font-semibold"
              style="color: rgb(74, 105, 226);"
            >
              <s>
                $
                125
              </s>
            </p>
          </div>
        </div>
        <div
          class="flex gap-[24px]"
        >
          <img
            alt="Add to favourite"
            class="cursor-pointer"
            data-nimg="1"
            decoding="async"
            height="32"
            loading="lazy"
            src="/heart-black.svg"
            style="color: transparent;"
            width="32"
          />
          <img
            alt="Delete Item"
            class="cursor-pointer"
            data-nimg="1"
            decoding="async"
            height="32"
            loading="lazy"
            src="/bin.svg"
            style="color: transparent;"
            width="32"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
